File: day7part1.go
package main
/*
# PROBLEM SPEC:
#    Given a list of input rules of the form:
#     <colour> bags contain <x> <colour> bags{, <y> <colour> bags}
#        OR
#     <colour> bags contain no other
#
#    ...Determine the number of <colour> bags that can contain at least one
#       Shiny Gold bag (either directly or indirectly)
*/
import (
  "fmt"
  "os"
  "bufio"
  "log"
  "strings"
  "strconv"
  "regexp"
)

type SubBagDef struct {
  colour string
  numbags int
}

type BagRule struct {
  colour string
  subbags []SubBagDef
}

func parseInputRulesFile(filename string) []BagRule {
  bagrules := []BagRule{}
  firstRE := regexp.MustCompile(`(?P<colour>.+) bags contain (?P<numbags>[0-9]+) (?P<subcolour>.+) bag`)
  noSubRE := regexp.MustCompile(`(?P<colour>.+) bags contain no other`)
  subRE   := regexp.MustCompile(`\D*(?P<numbags>[0-9]+) (?P<subcolour>.+) bag`)
  fh,err := os.Open(filename)
  if err != nil {
    log.Fatal(err)
  }
  defer fh.Close()
  s := bufio.NewScanner(fh)
  for s.Scan() {
    l := s.Text()
    //first level split: by comma. Check first element matches expectation
    elem := strings.Split(l,", ")
    //get match
    complexMatch := firstRE.FindAllStringSubmatch(elem[0],-1)
    if complexMatch == nil {
      terminalMatch := noSubRE.FindAllStringSubmatch(elem[0],-1)
      if terminalMatch == nil {
        log.Output(0,"SRC LINE INVALID: "+l)
      } else {
        bagrules = append(bagrules, BagRule{ colour: terminalMatch[0][1]})
      }
    } else {
      numBags,_ := strconv.Atoi(complexMatch[0][2])
      firstSubBag := SubBagDef{ colour: complexMatch[0][3], numbags: numBags}
      br := BagRule{ colour: complexMatch[0][1], subbags: []SubBagDef{firstSubBag}}
      for i:= 1; i<len(elem); i++ {
        sbr := subRE.FindAllStringSubmatch(elem[i],-1)
        if sbr != nil {
          sbc,_ := strconv.Atoi(sbr[0][1])
          sbd := SubBagDef{ colour: sbr[0][2], numbags: sbc}
          br.subbags = append(br.subbags,sbd)
        } else {
          log.Output(1, "SUB CLAUSE NO MATCH: "+elem[i])
        }
      }
      bagrules = append(bagrules,br)
    }
  }
  return bagrules
}

// ---------------------------------------------------------------------------
func findIsContainedBy(tgtColour string, bagrules []BagRule) []string {
  var isContainedBy []string
  for i := range bagrules {
    for j := range bagrules[i].subbags {
      if bagrules[i].subbags[j].colour == tgtColour {
        isContainedBy = append(isContainedBy, bagrules[i].colour)
      }
    }
  }
  return isContainedBy
}

// ---------------------------------------------------------------------------
func isInList(tgtColour string, listOfColours []string) bool {
  for i := range listOfColours {
    if listOfColours[i] == tgtColour {
      return true
    }
  }
  return false
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

func main() {
  bagrules := parseInputRulesFile("day7_input.txt")
  //bagrules := parseInputRulesFile("d7_test.txt")
  //fmt.Printf("%#v\n",bagrules)
  fmt.Println("Parsed " + strconv.Itoa(len(bagrules)) + " rules from input file")

  tgtColour := "shiny gold"
  alreadySeen := findIsContainedBy(tgtColour, bagrules)

  fmt.Println(tgtColour + " is directly contained by " + strconv.Itoa(len(alreadySeen)) + " bags: ")
  //fmt.Printf("%#v\n",alreadySeen)

  totCount := len(alreadySeen)
  //recurse and find sub-counts discarding those we've already seen
  searchList := alreadySeen
  for len(searchList)>0 {
    //Pop first element from the search list:
    inCol := searchList[0]
    searchList = searchList[1:]
    //find what colours contain this
    subContainsList := findIsContainedBy(inCol,bagrules)
    //If it's not already on our list, add it:
    for fCol := range subContainsList {
        if ! isInList(subContainsList[fCol],alreadySeen) {
          totCount += 1
          //stop double-counts
          alreadySeen = append(alreadySeen,subContainsList[fCol])
          //but search found colour for further indirection:
          searchList = append(searchList,subContainsList[fCol])
        }
    }
  }

  fmt.Println(tgtColour + " is directly and indirectly contained by " + strconv.Itoa(totCount) + " bag colours")
}

File: python/day7part1.py
#!/usr/bin/python3
#
# My working for Day 7, part 1
#
# PROBLEM SPEC:
#    Given a list of input rules of the form:
#     <colour> bags contain <x> <colour> bags{, <y> <colour> bags}
#
#    ...Determine the number of <colour> bags that can contain at least one
#       Shiny Gold bag (either directly or indirectly)
import sys
import re

#Input rules match one of two forms:
simprule = re.compile('(.+) bags contain ([0-9]+) (.+) bag')
comprule = re.compile('\D+([0-9]+) (.+) bag')
nocontains = re.compile('(.+) bags contain no other')

#Read the list of rules in from file
bagrules = {}
rfile="../data/day7_input.txt"
#rfile="d7_test.txt"
with open(rfile,"r") as rf:
    for rtext in rf:
        rule = {}
        contains = {}
        rparts=rtext.split(",")
        #First entry MUST exist and is the simple rule
        m = re.match(simprule,rparts[0])
        if m :
            col=m.group(1)
            contains[m.group(3)] = m.group(2)
            for rpart in rparts[1:len(rparts)]:
                o = re.match(comprule,rpart)
                if o:
                    contains[o.group(2)] = o.group(1)
        else :
            m = re.match(nocontains,rparts[0])
            if m :
                col=m.group(1)
            else :
                print("FAILED MATCH: {0}".format(rtext))
        bagrules[col] = contains

print("{0} Bag-contains rules found".format(len(bagrules)))

#Strictly speaking we now wish to invert the built structure to find the
#set of bags which can contain, directly or indirectly, our target "shiny gold"
#bag.

def colIsContainedBy(bagrules,desiredCol) :
    isContainedBy = []
    for col in bagrules.keys() :
        for bc in bagrules[col].keys() :
            if bc == desiredCol :
                isContainedBy.append(col)
    return isContainedBy

directContains = colIsContainedBy(bagrules,"shiny gold")

contCount = len(directContains)
print("{} direct contains:{}".format(contCount,directContains))
#Now we have a bunch of levels of indirection to look for.
alreadycounted = { i : 1 for i in directContains }
searchlist = directContains
while len(searchlist) > 0 :
    ccol=searchlist.pop()
    indCont = colIsContainedBy(bagrules,ccol)
    for iC in indCont :
        if iC not in alreadycounted :
            contCount = contCount + 1
            searchlist.append(iC)
            alreadycounted[iC] = 1

print("Recursive Search found a total of {0} direct/indirect contains".format(contCount))
